### ISSUE TITLE
Processing algorithm "Join by location" on selected features is way slower than usually

### ISSUE THREAD

**Issue by tschmetzer** [qgis/QGIS#44890]
### What is the bug or the crash?

Using the processing algorithm "Join by location" on selected features out of a larger layer is way slower than if the selected features are exported into layers. On the selection the processing requires more than an hour whereas on the layers it is a matter of 3 seconds. Note that the processing takes place on the same data and has the same results.

![2021-08-30 12_52_41-Mellilla_join_selection_by_location_ausschnitt](https://user-images.githubusercontent.com/67278094/131328882-a92ecd81-03f7-45aa-8a7f-5f85492edac6.png)

### Steps to reproduce the issue

1. Load layers
  a. Pop_Density_per_sqkm_Mollweide_Points_groesser0.sqlite (indexed, 3.5 GB, 25.965.154 point features)
  https://drive.google.com/file/d/1AJffraneCiLTq9oF041M3rQADUzdLG6x/view?usp=sharing (550 MB)
  b.  GHS_SMOD_groesser0_raster2points_ohneNoData.gpkg (indexed, 13.6 GB, 145.665.553 point features)
  https://drive.google.com/file/d/1TP5EwCew2K2ecOJHSqPfKgAK_i7rs6sO/view?usp=sharing (450 MB)
2. Graphically select any ~1.600 point features from layer Pop_Density_per_sqkm_Mollweide_Points_groesser0 and ~6.000 point features from GHS_SMOD_groesser0_raster2points_ohneNoData
![Mellilla_selection_Pop](https://user-images.githubusercontent.com/67278094/131337688-2aebbad3-a805-4291-b3f9-377771a9ea70.png)
![Mellilla_selection_SMOD](https://user-images.githubusercontent.com/67278094/131337712-f4b0c9b9-aaf4-4b6b-96ae-a6283a820648.png)

3. Save the selected features of each layer for later comparison ( I called them Melilla as I selected data around the area of the Spanish town Melilla in North Africa)
![smod_save_selected](https://user-images.githubusercontent.com/67278094/131337977-e0a640d0-c930-4387-8dcf-574750d5ede3.png)

4. Run the "Join by location" processing algorithm on the selection of the *original* layers Pop_Density_per_sqkm_Mollweide_Points_groesser0 and GHS_SMOD_groesser0_raster2points_ohneNoData according to the following screenshot
![Mellilla_join_selection_by_location](https://user-images.githubusercontent.com/67278094/131338266-557325af-fab6-4556-b439-429d031ce0b7.png)

5. Observe it will take a long time
![Mellilla_join_selection_by_location_forever](https://user-images.githubusercontent.com/67278094/131338313-c76ab042-f353-4826-9252-b91bf86d1a2a.png)

6. Compare to the execution time of the previously saved layers of the selection
![Mellilla_layers_join_by_location](https://user-images.githubusercontent.com/67278094/131338453-9575518a-a342-48f4-8a14-1ba6488d1650.png)
![Mellilla_layers_join_by_location_results](https://user-images.githubusercontent.com/67278094/131339059-733db68b-1a11-4b7c-aaa1-d3d13da1ef2a.png)

### Versions

QGIS version
3.20.2-Odense
QGIS code revision
9f59a156
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.3.1
PROJ version
8.1.0
EPSG Registry database version
v10.027 (2021-06-17)
GEOS version
3.9.1-CAPI-1.14.2
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 2009

Active Python plugins
GroupStats
QuickOSM
db_manager
MetaSearch
processing

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by tschmetzer** [comment]
Could it be the case that processing on selections doesn't make use of the index?

**Comment by gioman** [comment]
> On the selection the processing requires more than an hour

@tschmetzer this is how long it took here on master/linux

`Execution completed in 136.11 seconds (2 minutes 16 seconds)`

**Comment by uclaros** [comment]
I've been facing the same issue but never took the time to document it.
Performance is hugely affected when a selection is used for the *base* layer, while a selection on the *join* layer makes no noticeable difference for me.
The more obvious way to test is to run the tool on a dataset as a reference, then select all *base* features, tick *selected features only* and rerun. Execution time should be the same but is greatly increased depending on the *base* layer size.

**Comment by tschmetzer** [comment]
> The more obvious way to test is to run the tool on a dataset as a reference, then select all _base_ features, tick _selected features only_ and rerun. Execution time should be the same but is greatly increased depending on the _base_ layer size.

Do you mean to use as join layer the same layer as the base layer? If not can you post a screenshot for reproduction?

**Comment by tschmetzer** [comment]
> > On the selection the processing requires more than an hour
> 
> @tschmetzer this is how long it took here on master/linux
> 
> `Execution completed in 136.11 seconds (2 minutes 16 seconds)`

That's a different magnitude and may vary depending on the setup (Processor, RAM, storage connection, etc.). In my case I was using a network storage. I tested it with having the data on a local hard drive now and I get 
`Execution completed in 361.71 seconds (6 minutes 2 seconds)`
I suggest to focus on looking at differences with and without selection as differences in hardware configuration may lead to significant differences in execution time.
@gioman Have you compared the processing of the selected data to processing an export of the selected data? The latter was way faster in my case.

**Comment by uclaros** [comment]
- Start by having 2 layers, `POINT` and `POLYGON`
- Select all features on both layers.
- Run join attributes by location using *intersection*:

### The next three runs should have equal execution time:
![image](https://user-images.githubusercontent.com/11358178/133085637-3b52b68c-9e9d-48c0-8675-9950f4ca4c39.png)
Execution completed in 1.92 seconds

![image](https://user-images.githubusercontent.com/11358178/133085676-b9181d01-231e-4b67-980e-068e8c7c15f1.png)
Execution completed in 342.30 seconds (5 minutes 42 seconds)

![image](https://user-images.githubusercontent.com/11358178/133085721-835bfee3-eee0-409b-970d-e72d6ca923fa.png)
Execution completed in 2.06 seconds


### The next three runs should have equal execution time:
![image](https://user-images.githubusercontent.com/11358178/133085876-ed250808-a579-4d10-8f9d-321634044d7f.png)
Execution completed in 2.85 seconds

![image](https://user-images.githubusercontent.com/11358178/133085978-6174b12c-a4d7-4c24-9af8-518aa0472910.png)
Execution completed in 3.00 seconds

![image](https://user-images.githubusercontent.com/11358178/133086032-c984e65b-df7b-4bd8-9b4c-a5a09548c7cb.png)
Way too long, didn't wait!


### The next three runs should have equal execution time:
![image](https://user-images.githubusercontent.com/11358178/133086185-54afbb62-600f-437f-b6d7-e4e4ab40bee2.png)
Execution completed in 4.75 seconds

![image](https://user-images.githubusercontent.com/11358178/133086349-229ae315-41bf-4124-9803-e7909c0bf038.png)
Way too long, didn't wait!

![image](https://user-images.githubusercontent.com/11358178/133086266-cbeabca7-d870-467b-974f-e174b6258374.png)
Execution completed in 4.93 seconds

**Comment by gioman** [comment]
>  Execution completed in 342.30 seconds (5 minutes 42 seconds)

this is kind of puzzling, as for the other cases the common thing seems to be the selection on the polygon layer.

**Comment by tschmetzer** [comment]
> > Execution completed in 342.30 seconds (5 minutes 42 seconds)
> 
> this is kind of puzzling, as for the other cases the common thing seems to be the selection on the polygon layer.

Indeed a thrilling case. 

@gioman What further feedback is required?
@uclaros Thanks for the depiction! Can you attach your sample data for the layers POINT and POLYGON so that the issue can easily be reproduced and debugged?

**Comment by gioman** [comment]
> @uclaros Thanks for the depiction! Can you attach your sample data for the layers POINT and POLYGON so that the issue can easily be reproduced and debugged?

@tschmetzer I guess is not really about a specific dataset.

**Comment by tschmetzer** [comment]
> > @uclaros Thanks for the depiction! Can you attach your sample data for the layers POINT and POLYGON so that the issue can easily be reproduced and debugged?
> 
> @tschmetzer I guess is not really about a specific dataset.

I know. I just thought this would make reproduction and debugging easier :) Never mind.

**Comment by uclaros** [comment]
> @tschmetzer I guess is not really about a specific dataset.

Indeed. Just run the *Random point in extent* algorithm. The more points you generate, the more apparent the issue.

**Comment by roya0045** [comment]
That's strange, I'm unsure if that's something inherent to QgsProcessingFeatureSource or the algorithm itself. My first guess would be an issue where the returned featurecount is wront and it just return the integral, thus ordering the layers the wrong way. But I'm not sure how much that is the problem.

I'm curious to know if maintaning the 'order' of the dataset and inverting it (by keeping more or less feature selected int he bigger dataset than in the smaller one) has an effect. The geometry type may also have something to do with it, unsure.

Optimisations steps may need to be reworked depending on the findings. I'm someone has some time to test these theories that would help. I can look into some options soon-ish if I don't forget.

**Comment by roya0045** [comment]
I tested with small datasets, in any cases using select to subset the dataset reduced time, analysis with polygons took more times (they have more pvertex so computations is taking longer as expected). I can't quite replicate this with my datasets on 3.22 on windows. If someone could provide a small dataset to replicate, that would be appreciated.
